### DISCUSSION TITLE
isDirty not triggering with setValue and async sized form array

### DISCUSSION THREAD

**Discussion by afestbaum** [react-hook-form/react-hook-form#6570] [Q&A]
Hey community,

I'm currently working on a form that populates multiple input boxes that should be driven by an array (content and number of boxes). The content and number of array elements arrives async from an API. To achieve this I'm using an useEffect hook to monitor the source array and upon change (usually only once) I register the additional form fields. If the rendered inputs use setValue(..., { shouldDirty: true }) the isDirty formstate is not updated. I suspect this is because I set the defaultValue to an empty array at the useForm hook - although I thought reset would take care of this then.

What is a working way to achieve my goal of tracking the isDirty state? Attached you can find a codesandbox that illustrates the issue. The async operation is simulated by a button click that triggers the state update instead of the API call.

https://codesandbox.io/s/react-hook-form-formstate-dirty-touched-submitted-forked-72sl8?file=/src/index.js

Strangely if I wrap the inputs in a <Controller /> it works as expected with the same setValue calls. The reason for not going with this solution is that my special Input components should be reusable and not specific to one Form. Thus I would not like to pass down the control object ... as with typescript the type changes.

Edit: With v7.12.2 it is working as expected ... with the latest it is not

**Comment by bluebill1049** [comment]
```
reset({
  arrayWithUnknownContent: arrayWithUnknownContent,
  otherImportantField: "this should not be an array"
}, { keepDefaultValues: true }); // keep the default values unChnaged?
```

what about the above keep the defaultValues

**Reply by afestbaum** [reply]
Hey Bill, thank you for your time. My goal is not to retain defaultValues, but rather to get isDirty to work for the newly registered fields. If you look at my codesandbox, click the simulation button and edit the last input; isDirty remains false although setValue was called with the shouldDirty option.

**Reply by alexfestbaum** [reply]
This is fixed in 7.15.4

Thank you!
